lib/livebook/hubs/enterprise_client.ex

```diff
@@ -0,0 +1,118 @@
+defmodule Livebook.Hubs.EnterpriseClient do
+  @moduledoc false
+  use GenServer
+
+  alias Livebook.Hubs.Enterprise
+  alias Livebook.WebSocket
+  alias Livebook.WebSocket.Server
+
+  defstruct [:server, :hub, reply: %{}]
+
+  @doc """
+  Connects the Enterprise client with WebSocket server.
+  """
+  @spec start_link(Enterprise.t()) :: GenServer.on_start()
+  def start_link(%Enterprise{} = enterprise) do
+    GenServer.start_link(__MODULE__, enterprise)
+  end
+
+  @doc """
+  Sends a Request.
+  """
+  @spec send_request(pid(), WebSocket.proto()) :: :ok
+  def send_request(pid, %_struct{} = data) do
+    GenServer.call(pid, {:request, data})
+  end
+
+  @doc """
+  Disconnects the Enterprise client with WebSocket server.
+  """
+  @spec disconnect(pid()) :: :ok
+  def disconnect(pid) do
+    GenServer.cast(pid, :disconnect)
+  end
+
+  @doc """
+  Subscribe to Enterprise Client events.
+
+  ## Messages
+
+    * `{:connect, :error, reason}`
+
+  """
+  @spec subscribe() :: :ok | {:error, {:already_registered, pid()}}
+  def subscribe do
+    Phoenix.PubSub.subscribe(Livebook.PubSub, "enterprise")
+  end
+
+  @doc """
+  Unsubscribes from `subscribe/0`.
+  """
+  @spec unsubscribe() :: :ok
+  def unsubscribe do
+    Phoenix.PubSub.unsubscribe(Livebook.PubSub, "enterprise")
+  end
+
+  @doc """
+  Notifies interested processes about Enterprise Client messages.
+
+  Broadcasts the given message under the `"enterprise"` topic.
+  """
+  @spec broadcast_message(any()) :: :ok
+  def broadcast_message(message) do
+    Phoenix.PubSub.broadcast(Livebook.PubSub, "enterprise", message)
+  end
+
+  ## GenServer callbacks
+
+  @impl true
+  def init(%Enterprise{url: url, token: token} = enterprise) do
+    headers = [{"X-Auth-Token", token}]
+
+    {:ok, pid} = Server.start_link(url, headers)
```

Comment:
This looks great. The remaining change now is to pass the current pid to Server:

```suggestion
    {:ok, pid} = Server.start_link(self(), url, headers)
```

And now, instead of the server doing the broadcasts, it would only send messages to the `server`. The server is then responsible for broadcasting. There is no broadcasting at the client/server level.

WDYT?

Reply:
I prefer this idea